### ISSUE TITLE
Implementation of Criminal Identification System_#Team_5

### ISSUE THREAD

**Issue by Jessica-Salome** [jessy0506/jessy0506#1]
lV.	IMPLEMENTATION OF THE CRIMINAL IDENTIFICATION SYSTEM

To implement FaceNet one can use various open-source models which are pre-trained or train a model on their own dataset. The most famous open-source FaceNet model is OpenFace which is trained using PyTorch another one is known as “FaceNet by David Sandberg” which is trained using TensorFlow. For the purpose of the Criminal Identification system “FaceNet by Hiroki Taniai,” a pretrained Keras model is used. This Keras model was trained  on “MS-celeb-1M dataset”. For the criminal Identification System, a “Criminals Dataset” is used which  contains  around 200 snapshots of most wanted  criminals  in  the world. These pictures are clicked at  various angles, poses  and illumination with/without headgear, sunglasses, accessories etc. Some pictures also have no  visible  faces. The data set is divided into two parts: a training set and a validation/test set. The step-wise implementation is-

A.	Face Detection
MTCNN was directly implemented using the MTCNN library by “ipazc/mtcnn project”. All images were loaded as  a NumPy array and converted to RGB values. Then an MTCNN face detector class was made to detect faces in the image. The output was a list of bounding boxes  where  length and width were identified from the x and  y-axis.  After faces from all images have been detected they are stored as a compressed file.

B.	Face Embeddings
Face embeddings need to be created  so  comparisons  with different vectors can be done. This is the step where the FaceNet model was used for creating embeddings. After loading the compressed file of detected faces the  pixel  values need to be standardized as it is required for FaceNet.The pre-trained Keras facenet model is  loaded.  Each face is enumerated to find its prediction  and  embedding from the train and test set. The embeddings were saved as a compressed NumPy array.

C.	Face Classification
In this part of the process, embeddings are classified  using machine learning models to be identified as one of the criminals. Before applying a classification, model vector normalization is applied so values are scaled. The  scikit  learn normalization library is used for this  purpose. Next,  the names of the criminals are converted from string to integer format. This is done using LabelEncoder of scikit learn. The classification model used is Linear  Support  Vector Machine as it is effective for differentiating between the face embeddings. The linear SVM model is fit on the training data.

D.	Plotting Faces
To visualize the working of this entire model a face from the compressed test set is picked. Then embeddings for this image are created. This face embedding is  used as input to  fit in the model and get predictions.
